syne_tune/results_callback.py

```diff
@@ -0,0 +1,158 @@
+        self.csv_file = None
+        self.save_results_at_frequency = None
+        self.add_wallclock_time = add_wallclock_time
+        self._final_results_composer = final_results_composer
+        self._store_final_results_intermediately = store_final_results_intermediately
+        self._start_time_stamp = None
+        self._tuner = None
+
+    def _set_time_fields(self, result: Dict[str, Any]):
+        """
+        Note that we only add wallclock time to the result if this has not
+        already been done (by the backend)
+        """
+        if self._start_time_stamp is not None and ST_TUNER_TIME not in result:
+            result[ST_TUNER_TIME] = perf_counter() - self._start_time_stamp
+
+    def on_trial_result(
+        self, trial: Trial, status: str, result: Dict[str, Any], decision: str
+    ):
+        assert (
+            self.save_results_at_frequency is not None
+        ), "on_tuning_start must always be called before on_trial_result."
+        result = copy.copy(result)
+        result[ST_DECISION] = decision
+        result[ST_STATUS] = status
+        result[ST_TRIAL_ID] = trial.trial_id
+
+        for key in trial.config:
+            result[f"config_{key}"] = trial.config[key]
+
+        self._set_time_fields(result)
+
+        self.results.append(result)
+
+        if self.csv_file is not None:
+            self.save_results_at_frequency()
+
+    def store_results(self):
+        """
+        Store current results into CSV file, of name
+        ``{tuner.tuner_path}/{ST_RESULTS_DATAFRAME_FILENAME}``.
+        """
+        if self.csv_file is not None:
+            self.dataframe().to_csv(self.csv_file, index=False)
+        if self._store_final_results_intermediately:
+            self._write_final_results()
+
+    def dataframe(self) -> pd.DataFrame:
+        return pd.DataFrame(self.results)
+
+    def on_tuning_start(self, tuner):
+        # we set the path of the csv file once the tuner is created since the path may change when the tuner is stop
+        # and resumed again on a different machine.
+        self.csv_file = str(tuner.tuner_path / ST_RESULTS_DATAFRAME_FILENAME)
+        # we only save results every ``results_update_frequency`` seconds as this operation
+        # may be expensive on remote storage.
+        self.save_results_at_frequency = RegularCallback(
+            lambda: self.store_results(),
+            call_seconds_frequency=tuner.results_update_interval,
+        )
+        if self.add_wallclock_time:
+            self._start_time_stamp = perf_counter()
+        if self._final_results_composer is not None:
+            self._tuner = tuner
+
+    def _write_final_results(self):
+        if self._final_results_composer is not None:
+            final_results = self._final_results_composer(self._tuner)
+            if final_results is not None:
+                path = self._tuner.tuner_path / ST_FINAL_RESULTS_FILENAME
+                dump_json_with_numpy(final_results, path)
```

Comment:
How does this new method relate to this existing way of getting the final hyperparameters: https://github.com/awslabs/syne-tune/blob/main/examples/launch_plot_results.py#L69

```
tuning_experiment = load_experiment(tuner.name)
print(tuning_experiment)
print(f"best result found: {tuning_experiment.best_config()}")
```

Is this `_final_results_composer` needed if it's already possible to call `.best_config()`? Could we not re-use what `.best_config()` is doing?

Reply:
Final results are not best parameters. They can be anything. In the example I have here, they are statistics recorded by the scheduler, so nothing that would even be in `results.csv.zip`.

In the next PR, I am using this to record performance statistics for checkpoint removal. Again, this is not stored in `results.csv`.

These final results are optional. If you can obtain your answer from the `results.csv`, you should do that. But often, this is not possible.

Reply:
You have a good point, I'll do that